client/minicomm/subprocess_delegator.cc

```diff
@@ -0,0 +1,298 @@
+      config_->SubprocessConfig();
+  if (config_proto.filename().empty()) {
+    LOG(ERROR) << "Subprocess not configured.";
+    return;
+  }
+  std::unique_lock<std::mutex> pid_lock(child_pid_mutex_);
+  // If there is already child, or we are shutting down, we do nothing.
+  if (child_pid_ != 0) {
+    return;
+  }
+  int stdin_fd[2];
+  int stdout_fd[2];
+  int stderr_fd[2];
+  pipe(stdin_fd);
+  pipe(stdout_fd);
+  pipe(stderr_fd);
+  child_pid_ = fork();
+  if (child_pid_ == 0) {
+    close(stdin_fd[1]);
+    close(stdout_fd[0]);
+    close(stderr_fd[0]);
+
+    vector<const char*> newargv;
+    newargv.reserve(config_proto.argv_size() + 2);
+    newargv.push_back(config_proto.filename().c_str());
+    for (const string& arg : config_proto.argv()) {
+      newargv.push_back(arg.c_str());
+    }
+    newargv.push_back(NULL);
+
+    vector<const char*> newenv;
+    newenv.reserve(config_proto.env_size() + 1);
+    for (const string& v : config_proto.env()) {
+      newenv.push_back(v.c_str());
+    }
+    newenv.push_back(NULL);
+
+    dup2(stdin_fd[0], STDIN_FILENO);
+    if (stdin_fd[0] != STDIN_FILENO) {
+      close(stdin_fd[0]);
+    }
+    dup2(stdout_fd[1], STDOUT_FILENO);
+    if (stdout_fd[1] != STDOUT_FILENO) {
+      close(stdout_fd[1]);
+    }
+    dup2(stderr_fd[1], STDERR_FILENO);
+    if (stderr_fd[1] != STDERR_FILENO) {
+      close(stderr_fd[1]);
+    }
+
+    if (execve(config_proto.filename().c_str(),
+               const_cast<char* const*>(&newargv[0]),
+               const_cast<char* const*>(&newenv[0]))) {
+      // Failed to Exec. Report through stderr, then die quickly and without
+      // side effects.
+      static const char kExecveMessage[] = "Child Unable to execve!\n";
+      write(STDERR_FILENO, kExecveMessage, sizeof(kExecveMessage) - 1);
+      while (1) {
+        _exit(1);
+      }
+    }
+  }
+  close(stdin_fd[0]);
+  close(stdout_fd[1]);
+  close(stderr_fd[1]);
+  {
+    std::unique_lock<std::mutex> write_lock(write_mutex_);
+    write_stream_.reset(new proto2::io::FileOutputStream(stdin_fd[1]));
+  }
+  {
+    std::unique_lock<std::mutex> read_lock(read_mutex_);
+    read_stream_.reset(new proto2::io::FileInputStream(stdout_fd[0]));
+  }
+  {
+    std::unique_lock<std::mutex> error_lock(error_mutex_);
+    error_stream_.reset(new proto2::io::FileInputStream(stderr_fd[0]));
+  }
+  pid_lock.unlock();
+  child_spawned_.notify_all();
+}
+
+void SubprocessDelegator::KillChildProcess() {
+  std::unique_lock<std::mutex> pid_lock(child_pid_mutex_);
+  // If there is no child, or if we are in the process of shutting down,
+  // do nothing.
+  if (child_pid_ <= 0) {
+    return;
+  }
+
+  kill(child_pid_, SIGTERM);
+  // Give it time to end gracefully, then make sure it is really dead.
+  std::this_thread::sleep_for(std::chrono::seconds(5));
+  kill(child_pid_, SIGKILL);
+  int child_status;
+  // Don't let it become a zombie.
+  waitpid(child_pid_, &child_status, 0);
```

Comment:
Maybe specify WNOHANG here and spin in a sleep loop to limit the length of time we wait?